cmd/gpu_nfdhook/labeler.go

```diff
@@ -0,0 +1,209 @@
+			return gpuNameList, errors.Wrap(err, "Can't read device folder")
+		}
+
+		for _, drmFile := range drmFiles {
+			if l.controlDeviceReg.MatchString(drmFile.Name()) {
+				//Skipping possible drm control node
+				continue
+			}
+			devPath := path.Join(l.devfsDir, drmFile.Name())
+			if _, err := os.Stat(devPath); err != nil {
+				continue
+			}
+
+			gpuNameList = append(gpuNameList, f.Name())
+			break
+		}
+	}
+
+	return gpuNameList, nil
+}
+
+// getMemoryValues reads the GPU memory amount from the system.
+func (l *labeler) getMemoryAmount( /*cardNum*/ string) uint64 {
+	// reading GPU local memory amount is not yet available in the driver,
+	// so just return the environment variable value
+	envValue := os.Getenv(memoryOverrideEnv)
+	if envValue != "" {
+		val, err := strconv.ParseUint(envValue, 10, 64)
+		if err == nil {
+			return val
+		}
+	}
+	return 0
+}
+
+// addNumericLabel creates a new label if one doesn't exist. Else the new value is added to the previous value.
+func (lm labelMap) addNumericLabel(labelName string, valueToAdd int64) {
+	value := int64(0)
+	if numstr, ok := lm[labelName]; ok {
+		_, _ = fmt.Sscanf(numstr, "%d", &value)
+	}
+	value += valueToAdd
+	lm[labelName] = strconv.FormatInt(value, 10)
+}
+
+// createCapabilityLabels creates labels from the gpu capability file under debugfs.
+func (l *labeler) createCapabilityLabels(cardNum string) {
+	// try to read the capabilities from the i915_capabilities file
+	file, err := os.Open(filepath.Join(l.debugfsDRIDir, cardNum, "i915_capabilities"))
+	if err != nil {
+		klog.V(3).Infof("Couldn't open file:%s\n", err.Error()) // debugfs is not stable, there is no need to spam with error level prints
+		return
+	}
+	defer file.Close()
+
+	// define strings to search from the file, and the actions to take in order to create labels from those strings (as funcs)
+	searchStringActionMap := map[string]func(string){
+		"platform: ": func(platformName string) {
+			l.labels.addNumericLabel(labelNamespace+"platform_"+platformName+".count", 1)
+			l.labels[labelNamespace+"platform_"+platformName+".present"] = "true"
+		},
+		"gen: ": func(genName string) {
+			l.labels[labelNamespace+"platform_gen"] = genName
+		},
+	}
+
+	// Finally, read the file, and try to find the matches. Perform actions and reduce the search map size as we proceed. Return at 0 size.
+	scanner := bufio.NewScanner(file)
+	for scanner.Scan() {
+		for searchString, action := range searchStringActionMap {
+			var stringValue string
+			n, _ := fmt.Sscanf(scanner.Text(), searchString+"%s", &stringValue)
+			if n > 0 {
+				action(stringValue)
+				delete(searchStringActionMap, searchString)
+				if len(searchStringActionMap) == 0 {
+					return
+				}
+				break
+			}
+		}
+	}
+}
+
+// createLabels is the main function of plugin labeler, it creates label-value pairs for the gpus.
+func (l *labeler) createLabels() error {
+	gpuNameList, err := l.scan()
+	if err != nil {
+		klog.Error(err.Error())
+		return err
```

Comment:
No need to log the error here. Just return it, then log it in `main()` with the traceback together with
```go
err := ...
if err != nil {
        klog.Errorf("%+v", err)
        os.Exit(1)
}
```
This way you'll know exactly where the error happened.